crates/cubecl-opt/src/analyses/base.rs

```diff
@@ -0,0 +1,62 @@
+use std::{any::Any, cell::RefCell, rc::Rc};
+
+use type_map::TypeMap;
+
+use crate::Optimizer;
+
+use super::{
+    dominance::{Dominators, PostDominators},
+    liveness::Liveness,
+    post_order::PostOrder,
+};
+
+/// An analysis used by optimization passes. Unlike optimization passes, analyses can have state
+/// and persist until they're invalidated.
+pub trait Analysis {
+    /// Perform the analysis for the current optimizer state and return the persistent analysis state
+    fn init(opt: &mut Optimizer) -> Self;
+}
+
+#[derive(Default, Clone, Debug)]
+pub struct Analyses {
+    cache: Rc<RefCell<TypeMap>>,
+}
```

Comment:
It was hard for me to separate `Analysis` and `Analyses` visually while scanning code 😅. Would we rename `Analyses` to something like `AnalysisCache` or `AnalysisStore`?

Reply:
I was thinking the same thing but couldn't come up with a better name, `AnalysisCache` is probably best.